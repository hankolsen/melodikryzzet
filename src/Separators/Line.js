import React from 'react';
import PropTypes from 'prop-types';

import { CELL_WIDTH } from '../config';

const Line = ({ position, locations, direction }) => {
  let [x1, x2, y1, y2] = [0, 0, CELL_WIDTH / 2, CELL_WIDTH / 2];
  const lineLength = 11;
  const strokeWidth = 4;
  if (direction === 'across') {
    y1 = ((position.y + 1) * CELL_WIDTH) - (CELL_WIDTH / 3);
    y2 = y1;
    x1 = (position.x + locations[0]) * CELL_WIDTH;
    x2 = x1 + lineLength;
  } else {
    x1 = ((position.x + 1) * CELL_WIDTH) - (CELL_WIDTH / 3);
    x2 = x1;
    y1 = (position.y + locations[0]) * CELL_WIDTH;
    y2 = y1 + lineLength;
  }
  return (
    <line x1={x1} y1={y1} x2={x2} y2={y2} strokeWidth={strokeWidth} stroke="black" />
  );
};

Line.propTypes = {
  position: PropTypes.shape({
    x: PropTypes.number,
    y: PropTypes.number,
  }).isRequired,
  locations: PropTypes.arrayOf(PropTypes.number).isRequired,
  direction: PropTypes.string.isRequired,
};

export default Line;
