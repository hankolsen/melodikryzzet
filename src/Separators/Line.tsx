import React from 'react';

import { CELL_WIDTH } from '../config';
import { Direction, Position } from '../Crossword/Crossword.types';

type Props = {
  position: Position;
  location: number;
  direction: Direction;
};

const Line = ({ position, location, direction }: Props) => {
  let [x1, x2, y1, y2] = [0, 0, CELL_WIDTH / 2, CELL_WIDTH / 2];
  const lineLength = 11;
  const strokeWidth = 4;
  if (direction === 'across') {
    x1 = (position.x + location) * (CELL_WIDTH + 1) - lineLength / 2;
    x2 = x1 + lineLength;
    y1 += position.y * (CELL_WIDTH + 1);
    y2 = y1;
  } else {
    x1 = position.x * (CELL_WIDTH + 1) + CELL_WIDTH / 2 + 1;
    x2 = x1;
    y1 = (position.y + location) * (CELL_WIDTH + 1) - lineLength / 2 + 1;
    y2 = y1 + lineLength;
  }
  return (
    <line
      x1={x1}
      y1={y1}
      x2={x2}
      y2={y2}
      strokeWidth={strokeWidth}
      stroke="black"
    />
  );
};

export default Line;
