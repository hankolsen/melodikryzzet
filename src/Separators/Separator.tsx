import React from 'react';
import { Direction, Position } from 'Crossword/Crossword.types';

import Line from './Line';

type Props = {
  direction: Direction;
  position: Position;
  separator: string;
  locations: number[];
  id: string;
};

const Separator = ({
  direction,
  position,
  separator,
  locations,
  id,
}: Props) => {
  switch (separator) {
    case ',':
      return (
        <>
          {locations.map((location) => (
            <Line
              key={`${id}-${location}`}
              position={position}
              location={location}
              direction={direction}
            />
          ))}
        </>
      );
    default:
      return null;
  }
};

export default Separator;
