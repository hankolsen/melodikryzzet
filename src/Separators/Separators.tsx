import React from 'react';
import { SeparatorType } from 'Crossword/Crossword.types';
import Separator from './Separator';

type Props = {
  separators?: SeparatorType[];
};

const Separators = ({ separators }: Props) => {
  if (!separators || separators.length === 0) {
    return null;
  }
  return (
    <>
      {separators.map(({ direction, position, separator, locations, id }) => (
        <Separator
          key={`${id}-${separator}`}
          id={id}
          direction={direction}
          position={position}
          separator={separator}
          locations={locations}
        />
      ))}
    </>
  );
};

export default Separators;
