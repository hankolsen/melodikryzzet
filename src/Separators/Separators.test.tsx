import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Direction, SeparatorType } from 'Crossword/Crossword.types';
import Separators from './Separators';

describe('Separators test', () => {
  it('should not render anything', () => {
    const { container } = render(
      <svg>
        <Separators />
      </svg>,
    );

    expect(container.childNodes.length).toBe(1);
  });

  it('should still not render anything', () => {
    const separators: SeparatorType[] = [];

    const { container } = render(
      <svg>
        <Separators separators={separators} />
      </svg>,
    );

    expect(container.childNodes.length).toBe(1);
  });

  it('should render a single separator', () => {
    const separators: SeparatorType[] = [
      {
        direction: Direction.across,
        separator: ',',
        position: { x: 3, y: 8 },
        locations: [2],
        id: '1',
      },
    ];

    const { container } = render(
      <svg>
        <Separators separators={separators} />
      </svg>,
    );

    expect(container.querySelector('line')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render two separators', () => {
    const separators: SeparatorType[] = [
      {
        direction: Direction.across,
        separator: ',',
        position: { x: 5, y: 2 },
        locations: [2],
        id: '1',
      },
      {
        direction: Direction.down,
        separator: ',',
        position: { x: 2, y: 6 },
        locations: [3],
        id: '2',
      },
    ];

    const { container } = render(
      <svg>
        <Separators separators={separators} />
      </svg>,
    );

    expect(container.querySelector('line')).toBeInTheDocument();
    expect(container.querySelectorAll('line').length).toBe(2);
    expect(container).toMatchSnapshot();
  });

  it('should render multiple separators', () => {
    const separators: SeparatorType[] = [
      {
        direction: Direction.across,
        separator: ',',
        position: { x: 5, y: 2 },
        locations: [2, 4, 7],
        id: '1',
      },
    ];

    const { container } = render(
      <svg>
        <Separators separators={separators} />
      </svg>,
    );

    expect(container.querySelector('line')).toBeInTheDocument();
    expect(container.querySelectorAll('line').length).toBe(3);
    expect(container).toMatchSnapshot();
  });
});
