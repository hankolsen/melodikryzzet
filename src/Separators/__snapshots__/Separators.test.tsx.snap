// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Separators test should render a single separator 1`] = `
<div>
  <svg>
    <line
      stroke="black"
      stroke-width="4"
      x1="154.5"
      x2="165.5"
      y1="271.5"
      y2="271.5"
    />
  </svg>
</div>
`;

exports[`Separators test should render multiple separators 1`] = `
<div>
  <svg>
    <line
      stroke="black"
      stroke-width="4"
      x1="218.5"
      x2="229.5"
      y1="79.5"
      y2="79.5"
    />
    <line
      stroke="black"
      stroke-width="4"
      x1="282.5"
      x2="293.5"
      y1="79.5"
      y2="79.5"
    />
    <line
      stroke="black"
      stroke-width="4"
      x1="378.5"
      x2="389.5"
      y1="79.5"
      y2="79.5"
    />
  </svg>
</div>
`;

exports[`Separators test should render two separators 1`] = `
<div>
  <svg>
    <line
      stroke="black"
      stroke-width="4"
      x1="218.5"
      x2="229.5"
      y1="79.5"
      y2="79.5"
    />
    <line
      stroke="black"
      stroke-width="4"
      x1="80.5"
      x2="80.5"
      y1="283.5"
      y2="294.5"
    />
  </svg>
</div>
`;
